pkg/gator/sync/verify/verify.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/open-policy-agent/frameworks/constraint/pkg/core/templates"
+	gkapis "github.com/open-policy-agent/gatekeeper/v3/apis"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/cachemanager/parser"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/gator/reader"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var scheme *runtime.Scheme
+
+func init() {
+	scheme = runtime.NewScheme()
+	err := cfapis.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+	err = gkapis.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func Verify(unstrucs []*unstructured.Unstructured, flagDiscoveryResults string) (map[string][]int, error) {
+	discoveryResults, err := reader.ReadDiscoveryResults(flagDiscoveryResults)
+	if err != nil {
+		return nil, fmt.Errorf("reading: %w", err)
+	}
+
+	templates := []*templates.ConstraintTemplate{}
+	syncedGVKs := map[schema.GroupVersionKind]struct{}{}
+
+	for _, obj := range unstrucs {
+		if isSyncSet(obj) {
+			syncSet, err := reader.ToSyncSet(scheme, obj)
+			if err != nil {
+				return nil, fmt.Errorf("converting unstructured %q to syncset: %w", obj.GetName(), err)
+			}
+			for _, gvkEntry := range syncSet.Spec.GVKs {
+				gvk := schema.GroupVersionKind{
+					Group:   gvkEntry.Group,
+					Version: gvkEntry.Version,
+					Kind:    gvkEntry.Kind,
+				}
+				if _, exists := discoveryResults[gvk]; exists || discoveryResults == nil {
+					syncedGVKs[gvk] = struct{}{}
+				}
+			}
+		}
+		if isConfig(obj) {
+			config, err := reader.ToConfig(scheme, obj)
+			if err != nil {
+				return nil, fmt.Errorf("converting unstructured %q to config: %w", obj.GetName(), err)
+			}
+			for _, syncOnlyEntry := range config.Spec.Sync.SyncOnly {
+				gvk := schema.GroupVersionKind{
+					Group:   syncOnlyEntry.Group,
+					Version: syncOnlyEntry.Version,
+					Kind:    syncOnlyEntry.Kind,
+				}
+				if _, exists := discoveryResults[gvk]; exists || discoveryResults == nil {
+					syncedGVKs[gvk] = struct{}{}
+				}
+			}
+		}
+		if isTemplate(obj) {
+			templ, err := reader.ToTemplate(scheme, obj)
+			if err != nil {
+				return nil, fmt.Errorf("converting unstructured %q to template: %w", obj.GetName(), err)
+			}
+			templates = append(templates, templ)
+		}
+	}
+
+	missingReqs := map[string][]int{}
+
+	for _, templ := range templates {
+		// Fetch syncrequirements from template
+		syncRequirements, err := parser.ReadSyncRequirements(templ)
+		if err != nil {
+			return nil, fmt.Errorf("reading sync requirements from template %q: %w", templ.GetName(), err)
+		}
+		for i, requirement := range syncRequirements {
+			requirementMet := false
+			for gvk := range requirement {
+				if _, exists := syncedGVKs[gvk]; exists {
+					requirementMet = true
+				}
+			}
+			if !requirementMet {
+				missingReqs[templ.Name] = append(missingReqs[templ.Name], i+1)
+			}
+		}
+	}
+	return missingReqs, nil
```

Comment:
If I understand the function interface, it will be passing back a map of template names to the index of the assertion in the template's `requires-sync-data` annotation.

I think we'd be better off handing back a more sophisticated data structure with more flexible data.

I like

```
// key is a template name
map[string][]GVKEquivalenceSet
```

This makes the method re-usable by other logic that might seek to compose it.  It also makes your life much easier when you want to print out:

```
Template requirements missing:
* k8srequirepepperonipizza:
   - italian/v1:pizzaneopoltana OR italian/v1:pizzaromana
   - american/v1:newyorkstyle OR american/v1:deepdish
```

Reply:
This will also save you from dealing with integers, which I generally try to leave to interview questions lol.